File: Java/src/main/java/LinkedList/MiddleOfLinkedList/MiddleOfLinkedList.java
package LinkedList.MiddleOfLinkedList;

public class MiddleOfLinkedList {

    public static void main(String [] args) {
        ListNode node1 = new ListNode(1);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(4);

        ListNode node5 = new ListNode(5);
        ListNode node6 = new ListNode(6);


        node1.next = node2;
        node2.next = node3;
        node3.next = node4;

        node4.next = node5;
        node5.next = node6;


        ListNode node = middleNode(node1);
        while(node != null) {
            System.out.println(node.val);
            node = node.next;
        }
    }

    public static ListNode middleNode(ListNode node) {
        int count = 0;

        // Step1 - If only one node. Return Node
        if (node.next == null)
            return node;

        // Setp2 - Create pointers to keep track of positions
        ListNode pointer1 = node;
        ListNode pointer2 = node.next.next;

        // Step3 - Traverse through node
        while (pointer1 != null) {


            // Step4 - Check if pointer2 and next is null. Break if they are
            if (pointer2 == null || pointer2.next == null)
                break;


            System.out.println("Count: " + count + "    Pointer1: " + pointer1.val + "      Pointer2: " + pointer2.val);

            // Update the positions.
            // p1 should move by 1
            // p2 should move by 2
            pointer1 = pointer1.next;
            pointer2 = pointer2.next.next;

            // Increase count by 2. Because we want to know the total size
            // If size of 5 ...then 5 % 2 = 1, hence we leave middle element
            // If size of 6  ... then 6 % 2 = 0, hence we take the middle.next node
            count = count + 2;

        }



        return count % 2 == 1 ? pointer1 : pointer1.next;
    }
}

File: Python/project/src/Trees/MiddleOfLinkedList/MiddleOfLinkedList.py
from ListNodeObject import ListNode

class Solution(object):
    def middleNode(self, node):

        # Step1 - Create SlowPointer and FastPointer
        currentPointer = node
        fastPointer = node

        # Step2 - Traverse LinkedList using fastPointer
        while fastPointer != None and fastPointer.next != None:
            print("Current Node:", currentPointer.val, "  pointerDouble: ", fastPointer.val)

            # Step3 - Update pointers
            fastPointer = fastPointer.next.next         # Move pointer by two
            currentPointer = currentPointer.next        # Move pointer by one


        return currentPointer




def main():
    node1 = ListNode(1)
    node2 = ListNode(2)
    node3 = ListNode(3)
    node4 = ListNode(4)
    node5 = ListNode(5)
    node6 = ListNode(6)

    node1.next = node2
    node2.next = node3
    node3.next = node4
    node4.next = node5
    # node5.next = node6

    solution = Solution()
    output = solution.middleNode(node1)
    print("Middle Node: ", output.val)

if __name__ == '__main__':
    main()

'''
Notes
- A bit tricky. Still needed some help

Solution
1. Create two pointers. Slow and fast pointer
- Slow pointer will only traverse the list by one
- Fast pointer will only traverse the list by two

2. Stop the loop until the fastpointer is null or fastpointer.next is null
3. Return currentPointer

'''